roles/jd-client/src/lib/channel_manager/upstream_message_handler.rs

```
@@ -0,0 +1,434 @@
+        SupportedChannelTypes::Extended
+    }
+    fn is_work_selection_enabled_for_server(&self) -> bool {
+        true
+    }
+
+    #[instrument(skip_all, fields(request_id = ?msg.request_id, channel_id = ?msg.channel_id), level = "trace")]
+    async fn handle_open_standard_mining_channel_success(
+        &mut self,
+        msg: OpenStandardMiningChannelSuccess<'_>,
+    ) -> Result<(), Error> {
+        // Can't have standard channel with the pool under JD. fallback, if you
+        // ever receive this from pool.
+        info!("Received OpenStandardMiningChannelSuccess from Upstream");
+        debug!("OpenStandardMiningChannelSuccess: {msg:?}");
+        _ = self
+            .channel_manager_channel
+            .status_sender
+            .send(Status {
+                state: State::UpstreamShutdownFallback(JDCError::Shutdown),
+            })
+            .await;
+        Ok(())
+    }
+
+    #[instrument(skip_all, fields(request_id = msg.request_id, channel_id = msg.channel_id), level = "trace")]
+    async fn handle_open_extended_mining_channel_success(
+        &mut self,
+        msg: OpenExtendedMiningChannelSuccess<'_>,
+    ) -> Result<(), Error> {
+        info!("Received OpenExtendedMiningChannelSuccess from Upstream");
+        debug!("OpenExtendedMiningChannelSuccess: {msg:?}");
+
+        let channel_state = self.channel_manager_data.super_safe_lock(|data| {
+            let (hashrate, min_extranonce_size) = data
+                .pending_downstream_requests
+                .first()
+                .map(|req| match req {
+                    Mining::OpenExtendedMiningChannel(m) => {
+                        (m.nominal_hash_rate, m.min_extranonce_size)
+                    }
+                    Mining::OpenStandardMiningChannel(m) => {
+                        (m.nominal_hash_rate, self.min_extranonce_size)
+                    }
+                    _ => (100_000.0, self.min_extranonce_size),
```

Shouldn't we return an error in this case?